File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque
def solution(arr):
    direction = [(0,1),(1,0),(0,-1),(-1,0)]
    ans = 0
    lx,ly = 0,0
    sx,sy = 0,0
    ex,ey = 0,0
    n = len(arr)
    m = len(arr[0])
    for i in range(n):
        for j in range(m):
            if arr[i][j] == 'L':
                lx,ly = i,j
            if arr[i][j] == 'S':
                sx,sy = i,j
            if arr[i][j] == 'E':
                ex,ey = i,j
    def bfs(x,y,t):
        target = 'L' if t == 0 else 'E'
        q = deque()
        visit = [[False]*m for _ in range(n)]
        visit[x][y] = True
        q.append((x,y,0))
        while q:
            cx,cy,step = q.popleft()
            if arr[cx][cy] == target:
                return step
            for i in range(4):
                nx,ny = cx+direction[i][0],cy+direction[i][1]
                if 0<=nx<n and 0<=ny<m and not visit[nx][ny] and arr[nx][ny] != 'X':
                    visit[nx][ny] = True
                    q.append((nx,ny,step+1))
        return 0
    tol = bfs(sx,sy,0)
    if tol == 0: return -1
    toe = bfs(lx,ly,1)
    if toe == 0: return -1
    return tol+toe

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;
class Solution {
    int[][] dir = {{0,1},{0,-1},{1,0},{-1,0}};
    int n;
    int m;
    char[][] arr;
    public int solution(String[] maps) {
        n = maps.length;
        m = maps[0].length();
        int sx = 0;
        int sy = 0;
        int lx = 0;
        int ly = 0;
        arr = new char[n][m];
        // arr 초기화
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (maps[i].charAt(j) == 'S') {
                    sx = i;
                    sy = j;
                }
                if (maps[i].charAt(j) == 'L') {
                    lx = i;
                    ly = j;
                }
                arr[i][j] = maps[i].charAt(j);
            }
        }
        int tol = bfs(sx,sy,'L');
        int toe = bfs(lx,ly,'E');
        System.out.println("tol: " + tol + ", toe: " + toe);
        if (tol == -1) return -1;
        if (toe == -1) return -1;
        return tol+toe;
    }
    
    public int bfs(int x, int y, char t) {
        Queue<int[]> q = new ArrayDeque<>();
        boolean[][] visit = new boolean[n][m];
        visit[x][y] = true;
        int cnt = 0;
        q.add(new int[] {x,y,cnt});
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int cx = cur[0];
            int cy = cur[1];
            int step = cur[2];
            if (arr[cx][cy] == t) return step;
            for (int[] d : dir) {
                int nx = cx + d[0];
                int ny = cy + d[1];
                if (0 <= nx && nx < n && 0 <= ny && ny < m) {
                    if (!visit[nx][ny] && arr[nx][ny] != 'X') {
                        visit[nx][ny] = true;
                        q.add(new int[] {nx,ny,step+1});
                    }
                }
            }
        }
        return -1;
    }
}